Add new PrivateUse1 DeviceType for non-public devices

yuhc opened the pull request:
Summary: The new PrivateUse1 DeviceType is associated with the PrivateUse1 DispatchKey, which can be used for non-public devices without introducing a new device type. Note that the stringified name of the PrivateUse1 device is "privateuseone".

Test Plan: All CI should pass.

Differential Revision: D35859437

bdhirsh commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -1496,7 +1496,8 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
       common_device_.type() == DeviceType::ORT  ||
-      common_device_.type() == DeviceType::HPU) return;
+      common_device_.type() == DeviceType::HPU  ||
+      common_device_.type() == DeviceType::PrivateUse1) return;
```
Hmm was something breaking that required you to add `PrivateUse1` to the list here? It seems suspicious to add it here, because `PrivateUse1` has no guarantees around its storage

I actually don't think that we want this here, because it implies that the `PrivateUse1` device is

yuhc replied:
This makes sense. The iteration of an on-device tensor can always be done "on the device"; on the host we can just convert it back to CPU tensor first. Will remove this.

bdhirsh reviewed (approved):
other than the last minor comment, I think this look ok to me.

cc @albanD on the new device type. I think that long term this is ok, since we'll re-use the `DeviceType::PrivateUse1` for open registration (and we can think about un-winding the "device=privateuseone" conversion later)

albanD reviewed (commented):
Yes that sounds ok to me. 
I might suggest using `priv1`? or even `priv` if we don't plan on adding more? But not very important.

yuhc commented:
> using priv1?

The stringified name can't contain numbers; but it's probably better to contain the number to align with the used dispatch key. So "privateuseone` is used as a compromise. But please feel free to change the name in the future when the device registration thing comes up. 
Thanks everyone for approving the change!

yuhc committed:
Add new PrivateUse1 DeviceType for non-public devices (#77208)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/77208

The new PrivateUse1 DeviceType is associated with the PrivateUse1 DispatchKey, which can be used for non-public devices without introducing a new device type. Note that the stringified name of the PrivateUse1 device is "privateuseone".

Test Plan: All CI should pass.

Reviewed By: scottxu0730, bdhirsh

Differential Revision: D35859437

fbshipit-source-id: 4b71cf64ddc522aa2827eb1f7a25b79431bd8e44

bigfootjon commented:
@pytorchbot merge this

(Initiating merge automatically since Phabricator Diff has merged)
